src/cpp/session/modules/clang/CodeCompletion.cpp

```diff
@@ -154,6 +169,237 @@ core::json::Object toJson(const CodeCompleteResult& result)
+   // gcc and clang accept the same command)
+   std::string includePathOutput;
+   {
+      core::system::ProcessResult result;
+      std::string cmd = compilerPath + " -E -x c++ - -v < " kDevNull;
+      
+      // TODO: this needs to execute with Rtools on the PATH
+      Error error = core::system::runCommand(cmd, processOptions, &result);
+      if (error)
+         LOG_ERROR(error);
+      else if (result.exitStatus != EXIT_SUCCESS)
+         LOG_ERROR_MESSAGE("Error retrieving system include paths: " + result.stdOut);
+      else
+         includePathOutput = string_utils::trimWhitespace(result.stdOut);
+   }
+   if (includePathOutput.empty())
+      return;
+   
+   // strip out the include paths from the output
+   std::string startString = "#include <...> search starts here:";
+   std::string endString   = "End of search list.";
+   
+   std::string::size_type startPos = includePathOutput.find(startString);
+   if (startPos == std::string::npos)
+      return;
+   
+   std::string::size_type endPos = includePathOutput.find(endString, startPos);
+   if (endPos == std::string::npos)
+      return;
+   
+   std::string includePathsString = string_utils::trimWhitespace(
+            string_utils::substring(includePathOutput,
+                                    startPos + startString.size(),
+                                    endPos));
+   
+   // TODO: framework paths can be included here; how should we handle?
+   
+   // Split on newlines, and trim whitespace for each
+   std::vector<std::string> includePaths = core::algorithm::split(includePathsString, "\n");
+   for (std::size_t i = 0, n = includePaths.size(); i < n; ++i)
+      includePaths[i] = string_utils::trimWhitespace(includePaths[i]);
+   
+   pIncludePaths->insert(
+            pIncludePaths->end(),
+            includePaths.begin(),
+            includePaths.end());
+}
+
+json::Object jsonHeaderCompletionResult(const std::string& name,
+                                        const std::string& source,
+                                        int completionType)
+{
+   json::Object completionJson;
+   completionJson["type"]       = completionType;
+   completionJson["typed_text"] = name;
+   
+   json::Array textJson;
+   json::Object objectJson;
+   objectJson["text"]    = name;
+   objectJson["comment"] = source;
+   textJson.push_back(objectJson);
+   completionJson["text"] = textJson;
+
+   return completionJson;
+}
+
+Error getSystemHeaderCompletions(const std::string& token,
+                                 const std::string& parentDir,
+                                 const FilePath& filePath,
+                                 const std::string& docId,
+                                 const core::json::JsonRpcRequest& request,
+                                 core::json::JsonRpcResponse* pResponse)
+{
+   std::vector<std::string> includePaths;
+   
+   // discover the system headers
+   discoverSystemIncludePaths(&includePaths);
```

Comment:
Good idea! We might need to invalidate the cache if we detect that the compiler has changed though (ie, if `CXX` has changed) but that may be overkill since that will be done quite rarely and restarting the session would fix it.